db/changes_view.go

```diff
@@ -25,46 +26,95 @@ type channelsViewRow struct {
+
+	var queryRow QueryChannelsRow
+	found := results.Next(&queryRow)
+	if !found {
+		return nil, false
+	}
+	entry := &LogEntry{
+		Sequence:     queryRow.Sequence,
+		DocID:        queryRow.Id,
+		RevID:        queryRow.Rev,
+		Flags:        queryRow.Flags,
+		TimeReceived: time.Now(),
+	}
+
+	if queryRow.RemovalRev != "" {
+		entry.RevID = queryRow.RemovalRev
+		if queryRow.RemovalDel {
+			entry.SetDeleted()
+		} else {
+			entry.SetRemoved()
+		}
+	}
+	return entry, true
+
+}
+
 // Queries the 'channels' view to get a range of sequences of a single channel as LogEntries.
 func (dbc *DatabaseContext) getChangesInChannelFromView(
 	channelName string, endSeq uint64, options ChangesOptions) (LogEntries, error) {
 	if dbc.Bucket == nil {
 		return nil, errors.New("No bucket available for channel view query")
 	}
 	start := time.Now()
-	// Query the view:
-	optMap := changesViewOptions(channelName, endSeq, options)
+
+	startSeq := options.Since.SafeSequence() + 1
+
+	usingViews := dbc.Options.UseViews
 
 	entries := make(LogEntries, 0)
 	activeEntryCount := 0
 
-	base.LogToR("Cache", "  Querying 'channels' view for %q (start=#%d, end=#%d, limit=%d)", base.UD(channelName), options.Since.SafeSequence()+1, endSeq, options.Limit)
+	base.LogToR("Cache", "  Querying 'channels' view for %q (start=#%d, end=#%d, limit=%d)", base.UD(channelName), startSeq, endSeq, options.Limit)
+
 	// Loop for active-only and limit handling.
-	// The set of changes we get back from the view applies the limit, but includes both active and non-active entries.  When retrieving changes w/ activeOnly=true and a limit,
+	// The set of changes we get back from the query applies the limit, but includes both active and non-active entries.  When retrieving changes w/ activeOnly=true and a limit,
 	// this means we may need multiple view calls to get a total of [limit] active entries.
 	for {
-		vres := channelsViewResult{}
-		err := dbc.Bucket.ViewCustom(DesignDocSyncGateway(), ViewChannels, optMap, &vres)
+
+		// Query the view or index
+		queryResults, err := dbc.QueryChannels(channelName, startSeq, endSeq, options.Limit)
 		if err != nil {
-			base.LogfR("Error from 'channels' view: %v", err)
 			return nil, err
-		} else if len(vres.Rows) == 0 {
-			if len(entries) > 0 {
-				break
-			}
-			base.LogToR("Cache", "    Got no rows from view for %q", base.UD(channelName))
-			return nil, nil
 		}
+		queryRowCount := 0
 
-		// Convert the output to LogEntries:
+		// Convert the output to LogEntries.  Channel query and view result rows have different structure, so need to unmarshal independently.
 		highSeq := uint64(0)
-		for _, row := range vres.Rows {
-			entry := &LogEntry{
-				Sequence:     uint64(row.Key[1].(float64)),
-				DocID:        row.ID,
-				RevID:        row.Value.Rev,
-				Flags:        row.Value.Flags,
-				TimeReceived: time.Now(),
+		for {
+			var entry *LogEntry
+			var found bool
+			if usingViews {
```

Comment:
what about changing this to:

```
entry, found = nextChannelEntry(usingViews, queryResults)
```

which would then call  `nextChannelViewEntry()` or `nextChannelQueryEntry()`.  I think it would make the code slightly easier to read.